File: TypeScript/src/model/Name.ts
export class Name {

    private readonly name: string;

    public static named(name: string): Name {
        return new Name(name);
    }

    private constructor(name: string) {
        this.name = name;
    }

    public isNamedLike(other: Name): boolean {
        return this.equalsIgnoreCase(other);
    }

    private equalsIgnoreCase(other: Name): boolean {
        return this.name.toLowerCase() === other.name.toLowerCase();
    }

    public equals(other: any): boolean {
        if (other == null || !(other instanceof Name)) {
            return false;
        }
        const that: Name = other;
        return this.name === that.name;
    }

    public toString(): string {
        return this.name;
    }
}

File: Java/src/main/java/minimal/model/Name.java
package minimal.model;

public final class Name {

    private final String name;

    public static Name named(String name) {
        return new Name(name);
    }

    private Name(String name) {
        this.name = name;
    }

    public boolean isNamedLike(Name other) {
        return this.equalsIgnoreCase(other);
    }

    private boolean equalsIgnoreCase(Name other) {
        return name.equalsIgnoreCase(other.name);
    }

    @Override
    public int hashCode() {
        return name.hashCode();
    }

    @Override
    public boolean equals(Object other) {
        if (other == null || !(other instanceof Name)) {
            return false;
        }
        Name that = (Name) other;
        return this.name.equals(that.name);
    }

    @Override
    public String toString() {
        return name;
    }

}
